core/modules/memman/MemMan.h

```diff
@@ -0,0 +1,184 @@
+
+class TableInfo {
+public:
+
+    std::string tableName;    //< Name of the table
+
+    enum class LockType {
+        NOLOCK,               //< Item should not be locked
+        MUSTLOCK,             //< Item must be locked or declare failure
+        FLEXIBLE              //< Item may be locked if possible
+    };
+
+    LockType theData;         //< Lock options for the table's data
+    LockType theIndex;        //< Lock options for the table's index, if any
+
+    //-----------------------------------------------------------------------------
+    //! Constructor
+    //!
+    //! @param  tabName   is the name of the table.
+    //! @param  optData   lock options for the table's data
+    //! @param  optIndex  lock options for the table's index
+    //-----------------------------------------------------------------------------
+
+    TableInfo(std::string const& tabName,
+              LockType optData=LockType::MUSTLOCK,
+              LockType optIndex=LockType::NOLOCK)
+             : tableName(tabName), theData(optData), theIndex(optIndex)
+             {}
+};
+
+//-----------------------------------------------------------------------------
+//! @brief The memory manager.
+//!
+//! The MemMan is an abstract class the defines the interface to the memory
+//! manager that is used to lock database chunks in memory.
+//-----------------------------------------------------------------------------
+
+class MemMan {
+public:
+
+    //-----------------------------------------------------------------------------
+    //! @brief Create a memory manager and initialize for processing.
+    //!
+    //! @param  maxBytes   - Maximum amount of memory that can be used
+    //! @param  dbPath     - Path to directory where the database resides
+    //!
+    //! @return !0: The pointer to the memory manager.
+    //! @return  0: A manager could not be created.
+    //-----------------------------------------------------------------------------
+
+    static MemMan* create(uint64_t maxBytes, std::string const& dbPath);
+
+    //-----------------------------------------------------------------------------
+    //! @brief Lock a set of tables in memory for a particular chunk.
+    //!
+    //! @param  tables - Reference to the tables to process.
+    //! @param  chunk  - The chunk number associated with the tables.
+    //!
+    //! @return =0     - Nothing was locked. The errno variable holds the
+    //!                  reason, as follows:
+    //!                  xxxxxx - filesystem or memory error
+    //!                  ENOENT - a chunk was missing
+    //!                  ENOMEM - insufficient memory to fully satisfy request
+    //! @return !0     - Is the resource handle associated with this request.
+    //-----------------------------------------------------------------------------
+
+    using Handle = uint64_t;
+
+    struct HandleType {
+           static const Handle INVALID=0;
```

Comment:
Not at all since that was 7 vs 4. So, I scooted it over by three. Fell into the xroot format where we try to make things readable.